diff --git a/apps/s_server.c b/apps/s_server.c
index 6f2a2ae..e0590c2 100644
--- a/apps/s_server.c
+++ b/apps/s_server.c
@@ -60,6 +60,17 @@ typedef unsigned int u_int;
 #endif
 #include "internal/sockets.h"
 
+// forward declaration of __gcov_flush()
+void __gcov_flush();
+
+// signal handler definition which flushes coverage data
+void signal_handler(int signum)
+{
+    fprintf(stderr, "received signal %d, dump code coverage data\n", signum);
+    __gcov_flush();
+    exit(1);
+}
+
 static int not_resumable_sess_cb(SSL *s, int is_forward_secure);
 static int sv_body(int s, int stype, int prot, unsigned char *context);
 static int www_body(int s, int stype, int prot, unsigned char *context);
@@ -1049,6 +1060,12 @@ int s_server_main(int argc, char *argv[])
     int max_early_data = -1, recv_max_early_data = -1;
     char *psksessf = NULL;
 
+    fprintf(stdout, "set a signal handler\n");
+    if (signal(SIGINT, signal_handler) == SIG_ERR) {
+        fprintf(stderr, "could not set a handler for signal\n");
+        exit(1);
+    }
+
     /* Init of few remaining global variables */
     local_argc = argc;
     local_argv = argv;
